### PR TITLE
`SetMeterProvider` might miss the delegation for instruments and registries

### LINKED ISSUE #5757: instruments creation may cause panic when setting meter provider
### Description

This bug was found in [CI after running `TestMeterConcurrentSafe` test](https://github.com/open-telemetry/opentelemetry-go/actions/runs/10619828368/job/29438251680) and was introduced in #5754, where creating instruments while calling `setDelegate` method.

This is because creating instruments can be locked in this line:

https://github.com/open-telemetry/opentelemetry-go/blob/e47618fc36af51d17ecdcc7299bbf706397e1cb1/internal/global/meter.go#L350

While `setDelegate` is cleaning `instruments` map:

https://github.com/open-telemetry/opentelemetry-go/blob/e47618fc36af51d17ecdcc7299bbf706397e1cb1/internal/global/meter.go#L144

Then, the remaining code in the instrument creation method would trigger panic after the mutex is unlocked:

https://github.com/open-telemetry/opentelemetry-go/blob/e47618fc36af51d17ecdcc7299bbf706397e1cb1/internal/global/meter.go#L360

### Steps To Reproduce

1. Run

```sh
go test -timeout 30s -run ^TestMeterConcurrentSafe$ go.opentelemetry.io/otel/internal/global -count 1000
```

### Expected behavior

No Panic

### Proposal

- Fix the panic by cleaning the map but not releasing the map as a quick fix. #5758
- This panic behavior shows the possibility that the `SetMeterProvider` would miss the delegation setting for instruments and registry. We also need to fix this. The simple solution could be to use mutex to lock the whole instrumentation creation. #5780
- Concurrent safe tests didn't serve the purpose well due to the lack of runs (we only ran it once in CI). We can run the tests for 100 times. #5759

### PR DESCRIPTION
**Pull request by XSAM**
Closes #5757

This PR fixes an issue where `SetMeterProvider` might miss the delegation for instruments and registries. This bug brings a concurrent issue that could possibly make instruments and registries unable to operate correctly, such as recording, after using the `SetMeterProvider` method. The data put on these instruments and registries might be lost.

### CODE REVIEW

File: internal/global/meter_test.go [hunk 1]
```diff
@@ -162,8 +168,9 @@ func testSetupAllInstrumentTypes(t *testing.T, m metric.Meter) (metric.Float64Co
 // This is to emulate a read from an exporter.
 func testCollect(t *testing.T, m metric.Meter) {
 	if tMeter, ok := m.(*meter); ok {
-		m, ok = tMeter.delegate.Load().(metric.Meter)
-		if !ok {
+		// This changes the input m to the delegate.
+		m = tMeter.delegate
```
Comment (XSAM) [inline internal/global/meter_test.go:172]: By the way, it took me one hour to realize that this line would change the input, `m`. 🤦‍♂️

**Review by dashpole — APPROVED** [review]

**Review by dmathieu — APPROVED** [review]

**Review by MrAlias — COMMENTED** [review]
This change means that every measurement through the global API now requires at a minimum ~100ns to acquire the lock. This is on the hot-path and a performance evaluation should have been provided prior to this change being accepted.

The original design was aware of the case where a recording was performed concurrently with `SetMeterProvider` and it was dropped. That was intentional. There is a trade-off of performance and it was originally decided that speed on the hot-path is more important than synchronicity.

Are there panics or race conditions this addresses? If not these changes should be reverted until it can be shown that the performance impact is not significant.

**Comment by XSAM** [comment]
> The original design was aware of the case where a recording was performed concurrently with SetMeterProvider and it was dropped. That was intentional. There is a trade-off of performance and it was originally decided that speed on the hot-path is more important than synchronicity.

The drop would happen when initiating the instrument, not recording. The recording to instruments that didn't `setDelegate` would be ignored even after `SetMeterProvider`.

Imagine users initiating instruments while setting global meter provider. Some instruments were not able to record anything during the entire runtime of the application. This looks like race conditions to me.

**Comment by dashpole** [comment]
@MrAlias it isn't every measurement, right?  It is every instrument instantiation?  Or am I misreading

**Comment by MrAlias** [comment]
It looks like I misread this change. Looks good. Please disregard.

**Pull request merged**